### PR TITLE
967 - one-way attribute

### PR DESCRIPTION
**Pull request by andreawyss**
@treshugart this needs update to doc.
I need some help writing unit tests for this new feature.
Please take a look at the code. Thank you

### CODE REVIEW

**Comment by treshugart** [comment]
@andreawyss lookin good. What aspect of the tests do you want help with? I was thinking that the tests could be co-located with the existing tests in https://github.com/skatejs/skatejs/blob/master/test/unit/lifecycle/properties.js#L189. I know that file is getting a bit large, but we could refactor it after if we feel the need. Thoughts?

File: src/util/prop-definition.js [hunk 1]
```diff
@@ -76,6 +85,8 @@ export default class PropDefinition {
           break;
       }
     });
+
+    this.attrTargetIsSource = this.attrTarget && (this.attrTarget === this.attrSource);
```
Comment (treshugart) [inline src/util/prop-definition.js]: Do we need to even have the first part of this check here (`this.attrTarget && `)?

File: src/util/prop-definition.js [hunk 2]
```diff
@@ -54,7 +57,13 @@ export default class PropDefinition {
       // Only accept documented options and perform minimal input validation.
       switch (option) {
         case 'attribute':
-          this.attrName = resolveAttrName(optVal, nameOrSymbol);
+          if (isObject(optVal)) {
+            this.attrSource = resolveAttrName(optVal.source, nameOrSymbol);
```
Comment (treshugart) [inline src/util/prop-definition.js]: Would be nice to descructure the `optVal` props we're accessing here:

```js
const { source, target } = optVal;
```

File: src/util/prop-definition.js [hunk 3]
```diff
@@ -54,7 +57,13 @@ export default class PropDefinition {
...
+            this.attrTarget = resolveAttrName(optVal.target, nameOrSymbol);
+          } else {
+            this.attrSource = resolveAttrName(optVal, nameOrSymbol);
+            this.attrTarget = this.attrSource;
```
Comment (treshugart) [inline src/util/prop-definition.js]: Wherever we can it would be nice to save references to the props we're accessing. Sort of a nit, though.

File: src/util/error.js [hunk 4]
```diff
@@ -1,11 +1,3 @@
...
-import root from 'window-or-global';
-
 export default function error (message) {
-  const { DEBUG } = root;
-  if (!isUndefined(DEBUG) && DEBUG) {
-    console.error(message);
-  } else {
-    throw new Error(message);
-  }
+  throw new Error(message);
```
Comment (treshugart) [inline src/util/error.js]: Why the change here? Don't we still want to error in dev mode?

Comment (andreawyss) [inline src/util/error.js]: This will always throw. 
I removed the condition and the DEBUG flag. 
It will never console.error(), it always throws now.

**Comment by andreawyss** [comment]
@treshugart this PR is now ready for merge. Please, review it in details.

File: src/util/error.js (hunk 4 above)
Comment (treshugart) [inline src/util/error.js]: I probably should have worded that differently. We'd always want it to throw in dev, but don't we want this to be a noop in prod (i.e. keep DEBUG)?

**Comment by treshugart** [comment]
I've restarted the builds. This happens a lot with the browser testing services where they just fail for cryptic reasons.

File: src/util/error.js (hunk 4 above)
Comment (andreawyss) [inline src/util/error.js]: I think it should also throw in prop.

Comment (treshugart) [inline src/util/error.js]: Ok, let's leave that for now. Can revisit when we use it for more things.

**Comment by treshugart** [comment]
Now getting errors in IE https://travis-ci.org/skatejs/skatejs/builds/183068651#L665

**Comment by andreawyss** [comment]
@treshugart ok this PR is now ready. It works again in IE too.
Please review the changes to the doc too.

File: src/util/prop-definition.js [hunk 5]
```diff
@@ -72,26 +94,28 @@ export default class PropDefinition {
           this[option] = optVal;
           break;
         default:
-          error(`${option} is not a valid option. Options are: attribute, initial, default, coerce, deserialize, serialize.`);
+          deprecated(`${option} is not a valid option. Options are: attribute, initial, default, coerce, deserialize, serialize.`);
```
Comment (andreawyss) [inline src/util/prop-definition.js]: @treshugart here is the deprecated warning when using undosumented options.

File: src/util/deprecated.js [hunk 6]
```diff
@@ -0,0 +1,5 @@
+export default function deprecated (message) {
+  // TODO start throwing in a future version
```
Comment (treshugart) [inline src/util/deprecated.js]: I would say instead that we would think about just removing the deprecation message along with the functionality. New, breaking code shouldn't contain errors for legacy code, I don't think.

File: src/api/component.js [hunk 7]
```diff
@@ -90,6 +91,10 @@ function createInitProps (Ctor) {
   };
 }
 
+function tagName (elem) {
+  return elem && elem.tagName ? elem.tagName.toLowerCase() : '';
```
Comment (treshugart) [inline src/api/component.js]: Should be able to rely on `elem` to always be there. You can also use `localName` instead of `tagName.toLowerCase()`.

File: src/util/prop-definition.js (hunk 5 above)
Comment (treshugart) [inline src/util/prop-definition.js]: I wouldn't deprecate at all without a path forward. I see this as pretty low priority in the long run. Maybe we can just leave the old behaviour?

File: src/util/deprecated.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+  // throw new Error(message);
+  console.warn(message);
```
Comment (andreawyss) [inline src/util/deprecated.js]: Ok. What is the best way to know if you are in Dev or prod?

File: src/api/component.js (hunk 7 above)
Comment (andreawyss) [inline src/api/component.js]: done

File: src/util/deprecated.js (hunk 6 above)
Comment (andreawyss) [inline src/util/deprecated.js]: done

**Comment by andreawyss** [comment]
@treshugart can we merge this?

File: test/unit/lifecycle/properties.js [hunk 9]
```diff
@@ -810,4 +810,132 @@ describe('lifecycle/property', () => {
       });
     });
   });
+
+  describe('attribute one-way', () => {
+    it('source is not updated', (done) => {
+      const safe = uniqueId();
+      define(safe, {
```
Comment (treshugart) [inline test/unit/lifecycle/properties.js]: You don't need to pass a name anymore, you can do an anonymous define() :)

File: src/util/deprecated.js [hunk 10]
```diff
@@ -0,0 +1,5 @@
+export default function deprecated (elem, oldUsage, newUsage) {
+  // TODO: add logic to only warn in dev mode
```
Comment (treshugart) [inline src/util/deprecated.js]: A common pattern seems to be `process.env.NODE_ENV`. Most libs tend to favour this and it gets built out of the dist files.

Comment (treshugart) [inline src/util/deprecated.js]: Happy to either leave this as is and raise an issue to address it, or to do it now. If nether then I'd rather remove it as it'd be unused. I'll leave it up to you.
